Lib/Src/CUartCom.cpp

```diff
@@ -10,96 +10,231 @@
+ * send and receive data
+ */
+CUartCom::CUartCom(const std::string name)
+    : IComChannel(name),
+      mp_huart(nullptr),
+      m_uart_de_pin()
+{
+}
+
+/**
+ * @brief Initalise UART communication object
  * @param p_huart Pointer to UART hardware control register structure.
  */
-CUartCom::CUartCom(UART_HandleTypeDef *p_huart) : mp_huart(p_huart)
+bool CUartCom::init(UART_HandleTypeDef *p_huart)
 {
-    // Check mp_huart is not null
-    if (!mp_huart)
-    {
-        // Hang application with a toggling ALARM Pin
-        while (1)
-        {
-            // TODO: cannot use alarm pin. No space for it due to one less pin
-            // available on this MCU. HAL_GPIO_TogglePin(ALARM_GPIO_Port,
-            // ALARM_Pin);
-            HAL_Delay(200);
-        }
-    }
+    return init(p_huart, nullptr, 100);
 }
 
 /**
- * @brief Construct UART communication object with half-duplex mode.
- *
+ * @brief Initalise UART communication object with half-duplex mode.
  * @param p_huart Pointer to UART hardware control register structure.
  * @param uart_de_port Pointer to GPIO port.
  * @param uart_de_pin GPIO pin.
  */
-CUartCom::CUartCom(UART_HandleTypeDef *p_huart,
-                   GPIO_TypeDef *uart_de_port,
-                   uint16_t uart_de_pin)
-    : mp_huart(p_huart),
-      m_uart_de_port(uart_de_port),
-      m_uart_de_pin(uart_de_pin)
+bool CUartCom::init(UART_HandleTypeDef *p_huart,
+                    GPIO_TypeDef *uart_de_port,
+                    uint16_t uart_de_pin)
 {
+    bool b_success = true;
     // Check mp_huart is not null
-    if (!mp_huart)
+    if (!p_huart)
+    {
+        Error_Handler();
+        b_success = false;
+        return b_success;
+    }
+    else
     {
-        // Hang application with a toggling ALARM Pin
-        while (1)
+        mp_huart = p_huart;
+        m_uart_de_pin.init(uart_de_port, uart_de_pin);
+    }
+    // Check this hardware is not assigned to another instance
+    for (uint8_t i = 0; i < s_uart_instances; i++)
+    {
+        if (p_huart->Instance == sp_UART[i]->mp_huart->Instance)
         {
-            // TODO: cannot hardcode stuff like this into classes.
-            //            HAL_GPIO_TogglePin(ALARM_GPIO_Port, ALARM_Pin);
-            HAL_Delay(200);
+            b_success = false;
+            return b_success;
         }
     }
+
+    // Assign hardware to instance
+    sp_UART[s_uart_instances] = this;
+    s_uart_instances++;
+
+    return b_success;
+}
+
+/**
+ * @brief Enables UART reception with interrupts
+ * @note UART Interrupts must be enabled in the hardware configuration in order
+ * to use this method
+ *
+ */
+void CUartCom::startRx()
+{
+    HAL_UART_Receive_IT(mp_huart, &m_rx_char, 1);
 }
 
 /**
  * @brief sends message via UART.
  * @param msg Message to send.
+ * @return True if there was space in the transmission queue
  */
-void CUartCom::sendMessage(const std::string &msg)
+void CUartCom::send(const std::string msg)
```

Comment:
What exactly does this method do? What happens when the msg is empty?

Reply:
But then I can't store messages in the TX queue if the UART is busy transmitting.

Reply:
I don't follow how storing messages in Tx queue is related to Rx interrupt?
When using half-duplex uart both incoming and outgoing messages go through the same set of wires. Some modern half-duplex transceivers prevent echo by blocking Rx when Tx is active, but not all of them do that. This is why Rx interrupt should be disabled when you are transmitting and when you resume receiving, whatever is in the receive uart should be flushed out.

Reply:
I was testing it by echoing what I was sending to the MCU, and when disabling the RX Interrupts when transmitting, some messages were not being stored in the RX Queue. You're right, maybe there could be a check to disable RX interrupts only when it is half-duplex? Because if the MCU is busy transmitting a bunch of data, then it will not be able to store incoming commands (using half-duplex).